buildpacks/ruby/src/layers/bundle_install_layer.rs

```diff
@@ -53,211 +36,103 @@ pub(crate) fn call(
+#[derive(Deserialize, Serialize, Debug, Clone, PartialEq, CacheDiff, TryMigrate)]
+#[cache_diff(custom = clear_v1)]
 #[try_migrate(from = None)]
-pub(crate) struct MetadataV1 {
-    pub(crate) stack: String,
-    pub(crate) ruby_version: ResolvedRubyVersion,
-    pub(crate) force_bundle_install_key: String,
-    pub(crate) digest: MetadataDigest, // Must be last for serde to be happy https://github.com/toml-rs/toml-rs/issues/142
-}
-
-#[derive(Deserialize, Serialize, Debug, Clone, Eq, PartialEq, TryMigrate)]
-#[try_migrate(from = MetadataV1)]
 #[serde(deny_unknown_fields)]
-pub(crate) struct MetadataV2 {
-    pub(crate) distro_name: String,
-    pub(crate) distro_version: String,
+pub(crate) struct MetadataV3 {
+    #[cache_diff(rename = "OS Distribution")]
+    pub(crate) os_distribution: OsDistribution,
+    #[cache_diff(rename = "CPU Architecture")]
     pub(crate) cpu_architecture: String,
+    #[cache_diff(rename = "Ruby version")]
     pub(crate) ruby_version: ResolvedRubyVersion,
+    #[cache_diff(ignore)]
     pub(crate) force_bundle_install_key: String,
-    pub(crate) digest: MetadataDigest, // Must be last for serde to be happy https://github.com/toml-rs/toml-rs/issues/142
+
+    #[cache_diff(ignore)]
+    // Placeholder for a deprecated struct
+    pub(crate) digest: toml::Value, // Must be last for serde to be happy https://github.com/toml-rs/toml-rs/issues/142
 }
 
+// Introduced to drop support for MetadataDigest
+// 2025-05-16
 #[derive(Deserialize, Serialize, Debug, Clone, Eq, PartialEq, CacheDiff, TryMigrate)]
-#[cache_diff(custom = clear_v1)]
-#[try_migrate(from = MetadataV2)]
+#[try_migrate(from = MetadataV3)]
 #[serde(deny_unknown_fields)]
-pub(crate) struct MetadataV3 {
+pub(crate) struct MetadataV4 {
     #[cache_diff(rename = "OS Distribution")]
     pub(crate) os_distribution: OsDistribution,
     #[cache_diff(rename = "CPU Architecture")]
     pub(crate) cpu_architecture: String,
     #[cache_diff(rename = "Ruby version")]
     pub(crate) ruby_version: ResolvedRubyVersion,
-    #[cache_diff(ignore)]
-    pub(crate) force_bundle_install_key: String,
-
-    /// A struct that holds the cryptographic hash of components that can
-    /// affect the result of `bundle install`. When these values do not
-    /// change between deployments we can skip re-running `bundle install` since
-    /// the outcome should not change.
-    ///
-    /// While a fully resolved `bundle install` is relatively fast, it's not
-    /// instantaneous. This check can save ~1 second on overall build time.
-    ///
-    /// This value is cached with metadata, so changing the struct
-    /// may cause metadata to be invalidated (and the cache cleared).
-    ///
-    #[cache_diff(ignore)]
-    pub(crate) digest: MetadataDigest, // Must be last for serde to be happy https://github.com/toml-rs/toml-rs/issues/142
 }
 
-fn clear_v1(_new: &Metadata, old: &Metadata) -> Vec<String> {
+fn clear_v1(_new: &MetadataV3, old: &MetadataV3) -> Vec<String> {
```

Comment:
Very minor - I found this method name slightly confusing in the context of multiple `MetadataV*` structs + migrations + diffing. My first thought was that this method maybe clears `MetadataV1` stuff from `MetadataV3` as part of a migration process but that's obviously not the case given the argument types. I had to jump through the code a bit to realize it's for the diff output. Maybe this function could be called something more descriptive?

Reply:
TBF I got confused with it too for the same reason in refactoring. So I agree. I updated in https://github.com/heroku/buildpacks-ruby/pull/425/commits/7025309dde0080b61444941ee8a6fcef17422047. This names things more in line with their behaviors than their intent. In theory, additional logic could go here without needing to rename the function, which is a property I like.

The only downside is if we needed different custom cache clear functions for different structs, then this is the "best" name so we would have to come up with some convention, but matching it to the struct version seems good.